go/filemanager/pfsmodules/chunkmeta.go

```diff
@@ -0,0 +1,209 @@
+	if p.ChunkSize < defaultMinChunkSize ||
+		p.ChunkSize > defaultMaxChunkSize {
+		return errors.New(StatusBadChunkSize)
+	}
+
+	return nil
+}
+
+// CloudCheck checks the conditions when running on cloud.
+func (p *ChunkMetaCmd) CloudCheck() error {
+	if !IsCloudPath(p.FilePath) {
+		return errors.New(StatusShouldBePfsPath + ": p.FilePath")
+	}
+	if !CheckUser(p.FilePath) {
+		return errors.New(StatusUnAuthorized + ":" + p.FilePath)
+	}
+	return p.checkChunkSize()
+}
+
+// LocalCheck checks the conditions when running locally.
+func (p *ChunkMetaCmd) LocalCheck() error {
+	return p.checkChunkSize()
+}
+
+// NewChunkMetaCmdFromURLParams get a new ChunkMetaCmd.
+func NewChunkMetaCmdFromURLParam(r *http.Request) (*ChunkMetaCmd, error) {
+	method := r.URL.Query().Get("method")
+	path := r.URL.Query().Get("path")
+	chunkStr := r.URL.Query().Get("chunksize")
+
+	if len(method) == 0 ||
+		method != ChunkMetaCmdName ||
+		len(path) == 0 ||
+		len(chunkStr) == 0 {
+		return nil, errors.New(http.StatusText(http.StatusBadRequest))
+	}
+
+	chunkSize, err := strconv.ParseInt(chunkStr, 10, 64)
+	if err != nil {
+		return nil, errors.New(StatusBadChunkSize)
+	}
+
+	return &ChunkMetaCmd{
+		Method:    method,
+		FilePath:  path,
+		ChunkSize: chunkSize,
+	}, nil
+}
+
+type metaSlice []ChunkMeta
+
+func (a metaSlice) Len() int           { return len(a) }
+func (a metaSlice) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a metaSlice) Less(i, j int) bool { return a[i].Offset < a[j].Offset }
+
+// GetDiffChunkMeta gets difference between srcMeta and dstMeta.
+func GetDiffChunkMeta(srcMeta []ChunkMeta, dstMeta []ChunkMeta) ([]ChunkMeta, error) {
+	if len(dstMeta) == 0 || len(srcMeta) == 0 {
+		return srcMeta, nil
+	}
+
+	if !sort.IsSorted(metaSlice(srcMeta)) {
+		sort.Sort(metaSlice(srcMeta))
+	}
+
+	if !sort.IsSorted(metaSlice(dstMeta)) {
+		sort.Sort(metaSlice(dstMeta))
+	}
+
+	dstIdx := 0
+	srcIdx := 0
+	diff := make([]ChunkMeta, 0, len(srcMeta))
+
+	for {
+		if srcMeta[srcIdx].Offset < dstMeta[dstIdx].Offset {
+			diff = append(diff, srcMeta[srcIdx])
+			srcIdx++
+		} else if srcMeta[srcIdx].Offset > dstMeta[dstIdx].Offset {
+			dstIdx++
+		} else {
+			if srcMeta[srcIdx].Checksum != dstMeta[dstIdx].Checksum {
+				diff = append(diff, srcMeta[srcIdx])
+			}
+
+			dstIdx++
+			srcIdx++
+		}
+
+		if dstIdx >= len(dstMeta) {
+			break
+		}
+
+		if srcIdx >= len(srcMeta) {
+			break
+		}
+	}
+
+	if srcIdx < len(srcMeta) {
+		diff = append(diff, srcMeta[srcIdx:]...)
+	}
+
+	return diff, nil
+}
+
+// GetChunkMeta gets chunk metas from path of file.
+func GetChunkMeta(path string, len int64) ([]ChunkMeta, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer Close(f)
+
+	if len > defaultMaxChunkSize || len < defaultMinChunkSize {
+		return nil, errors.New(StatusBadChunkSize)
+	}
+
+	fi, err := f.Stat()
+	if err != nil {
+		return nil, err
+	}
+
+	metas := make([]ChunkMeta, 0, fi.Size()/len+1)
```

Comment:
Sorry I don't understand why there is a `+1` in `fi.Size()/len+1`, I thought `(fi.Size() + len - 1)/len` is the `ceil(fi.Size()/len)`.
Anyway, I think `var metas []ChunkMeta` is more fool proof, and easy to read. The optimization here will not have any visible impact to the performance. We should choose code clarity rather than performance, unless performance is a bottleneck.